include/dicom/dicom.h

```diff
@@ -1479,91 +1635,167 @@ DcmFilehandle *dcm_filehandle_create_from_memory(DcmError **error,
+const DcmDataSet *dcm_filehandle_get_file_meta(DcmError **error,
+                                               DcmFilehandle *filehandle);
+
+/**
+ * Get Transfer Syntax UID for a fileahndle.
+ *
+ * :param filehandle: File
+ *
+ * :return: UID of the transfer syntax for this File.
+ */
+DCM_EXTERN
+const char *dcm_filehandle_get_transfer_syntax_uid(const DcmFilehandle *filehandle);
 
 /**
  * Read metadata from a File.
  *
- * Keeps track of the offset of the Pixel Data Element relative to the
- * beginning of the filehandle to speed up subsequent access to individual
- * Frame items.
+ * Read slide metadata, stopping when one of the tags in the stop list is
+ * seen. If the stop list pointer is NULL, it will stop on any of the pixel
+ * data tags.
+ *
+ * The return result must be destroyed with :c:func:`dcm_dataset_destroy()`.
+ *
+ * After calling this function, the filehandle read point is always
+ * positioned at the tag that stopped the read. You can call this function
+ * again with a different stop set to read more of the metadata.
  *
  * :param error: Pointer to error object
  * :param filehandle: File
+ * :param stop_tags: NULL, or Zero-terminated array of tags to stop on
  *
  * :return: metadata
  */
 DCM_EXTERN
 DcmDataSet *dcm_filehandle_read_metadata(DcmError **error,
-                                         DcmFilehandle *filehandle);
+                                         DcmFilehandle *filehandle,
+                                         const uint32_t *stop_tags);
 
 /**
- * Read Basic Offset Table from a File.
+ * Get metadata from a File.
+ *
+ * Gets the File's metadata and saves it in the File handle. Returns a
+ * reference to this internal copy of the File metadata.
+ *
+ * The return result must not be destroyed. Make a clone of it with
+ * :c:func:`dcm_dataset_clone()` if you need it to remain valid after
+ * closing the File handle.
+ *
+ * After calling this function, the filehandle read point is always
+ * positioned at the tag that stopped the read.
  *
- * In case the Pixel Data Element does not contain a Basic Offset Table item,
- * but contains an Extended Offset Table element, the value of the Extended
- * Offset Table element will be read instead.
+ * It is safe to call this function many times.
  *
  * :param error: Pointer to error object
  * :param filehandle: File
- * :param metadata: Metadata
  *
- * :return: Basic Offset Table
+ * :return: metadata
  */
 DCM_EXTERN
-DcmBOT *dcm_filehandle_read_bot(DcmError **error, DcmFilehandle *filehandle,
-                                DcmDataSet *metadata);
+const DcmDataSet *dcm_filehandle_get_metadata(DcmError **error,
+                                              DcmFilehandle *filehandle);
 
 /**
- * Build Basic Offset Table for a File.
+ * Read everything necessary to fetch frames from the file.
+ *
+ * Scans the PixelData sequence and loads the PerFrameFunctionalGroupSequence,
+ * if present.
```

Comment:
The Per-Frame Functional Group Sequence is not part of the Pixel Data element.

Unless the function reads the entire pixel data, I suggest renaming it. Maybe `dcm_filehandle_seek_pixeldata`?

Reply:
It scans pixeldata and finds all the offsets, and also loads in per frame functional group sequence, if present.

Reply:
The name `dcm_filehandle_read_pixeldata` implies that the function reads the entire pixel data elements and loads the data into memory.

> It scans pixeldata and finds all the offsets

I suggest naming the function `dcm_filehandle_find_frame_offsets`, `dcm_filehandle_build_frame_offset_table`, or something along those lines.

> also loads in per frame functional group sequence, if present

I understand the motivation to group reading of the per-frame functional groups items with reading the frame offset items. However, conceptually they are two different things since the latter is part of the pixel data element and the former are metadata elements. Therefore, I would keep them separate.

Reply:
In my mind it was something like "get ready to read pixeldata", so it fetches all the parts that will be required for that.

Maybe `dcm_filehandle_prepare_read_frame()`?

Reply:
I've renamed this.

Reply:
What bothers me about this function is that its behavior is opaque and that it introduces state.

Previously, the caller needed to get the (complete) metadata and the frame offset table prior to reading a frame and that information needed to get passed to the `read_frame` function. Now, a number of functions need to be called in the right order and it's not clear why this is the case and what's happening under the hood.

Thoughts?

Reply:
You can call the functions in any order, and they are all optional. Although there is internal state (necessary for performance), the user doesn't need to know about it.

`dcm_filehandle_prepare_read_frame()` is only split out as a separate function for programs which want to control the delay you get with the first `read_frame`.